adr/2025_06_17-formats.md

```diff
@@ -0,0 +1,59 @@
+# BlockNote Formats
+
+Right now, there are several formats supported by BlockNote:
+
+- internal html (serialized html from the editor's schema) for things like clipboard handling
+- external html (a "normal" html format) for things like rendering to a blog
+- markdown
+- exported formats: docx, pdf, react
+
+There is a relationship between all of these formats:
+
+```ts
+/** The editor content */
+type Model = Block[] | InternalEditorState;
+
+type InternalHTML = string;
+function toInternalHTML(model: Model): InternalHTML;
+function fromInternalHTML(html: InternalHTML): Model;
+
+type ExternalHTML = string;
+function toExternalHTML(model: Model): ExternalHTML;
+function fromExternalHTML(html: ExternalHTML): Model;
+
+type RenderedHTML = string; // What renders to the editor in the DOM
+function toRenderedHTML(model: Model): RenderedHTML;
+function fromRenderedHTML(html: RenderedHTML): Model;
+
+type Markdown = string;
+function toMarkdown(model: ExternalHTML): Markdown;
+function fromMarkdown(markdown: Markdown): Model;
+
+type Docx = Buffer;
+function toDocx(model: Model): Docx;
+
+type PDF = Buffer;
+function toPDF(model: Model): PDF;
+```
+
+You'll notice that the formats are all derived from the model.
+
+This gives us a unidirectional flow of data. Any other direction is potentially lossy (e.g. markdown -> model).
```

Comment:
Isn't `Model -> ***` also potentially lossy if the target format doesn't have mappings to aspects of the model (e.g. nested blocks in Markdown)?